Userland/Libraries/LibPthread/semaphore.cpp

```diff
@@ -64,116 +46,110 @@ int sem_init(sem_t* sem, int shared, unsigned int value)
         return -1;
     }
 
-    auto rc = pthread_mutex_init(&sem->mtx, nullptr);
-    if (rc != 0) {
-        errno = rc;
-        return -1;
-    }
-
-    rc = pthread_cond_init(&sem->cv, nullptr);
-    if (rc != 0) {
-        errno = rc;
-        return -1;
-    }
-
     sem->value = value;
-
     return 0;
 }
 
-sem_t* sem_open(const char*, int, ...)
+int sem_destroy(sem_t*)
 {
-    errno = ENOSYS;
-    return nullptr;
+    return 0;
 }
 
-int sem_post(sem_t* sem)
+int sem_getvalue(sem_t* sem, int* sval)
 {
-    auto rc = pthread_mutex_lock(&sem->mtx);
-    if (rc != 0) {
-        errno = rc;
-        return -1;
-    }
-
-    if (sem->value == SEM_VALUE_MAX) {
-        pthread_mutex_unlock(&sem->mtx);
-        errno = EOVERFLOW;
-        return -1;
-    }
-
-    sem->value++;
-
-    rc = pthread_cond_signal(&sem->cv);
-    if (rc != 0) {
-        pthread_mutex_unlock(&sem->mtx);
-        errno = rc;
-        return -1;
-    }
-
-    rc = pthread_mutex_unlock(&sem->mtx);
-    if (rc != 0) {
-        errno = rc;
-        return -1;
-    }
-
+    u32 value = AK::atomic_load(&sem->value, AK::memory_order_relaxed);
+    *sval = value & ~POST_WAKES;
     return 0;
 }
 
-int sem_trywait(sem_t* sem)
+int sem_post(sem_t* sem)
 {
-    auto rc = pthread_mutex_lock(&sem->mtx);
-    if (rc != 0) {
-        errno = rc;
-        return -1;
-    }
-
-    if (sem->value == 0) {
-        pthread_mutex_unlock(&sem->mtx);
-        errno = EAGAIN;
-        return -1;
-    }
-
-    sem->value--;
-
-    rc = pthread_mutex_unlock(&sem->mtx);
-    if (rc != 0) {
-        errno = rc;
-        return -1;
-    }
+    u32 value = AK::atomic_fetch_add(&sem->value, 1u, AK::memory_order_release);
+    // Fast path: no need to wake.
+    if (!(value & POST_WAKES)) [[likely]]
+        return 0;
 
+    // Pass the responsibility for waking more threads if more slots become
+    // available later to sem_wait() in the thread we're about to wake, as
+    // opposed to further sem_post() calls that free up those slots.
+    value = AK::atomic_fetch_and(&sem->value, ~POST_WAKES, AK::memory_order_relaxed);
+    // Check if another sem_post() call has handled it already.
+    if (!(value & POST_WAKES)) [[likely]]
+        return 0;
+    int rc = futex_wake(&sem->value, 1);
+    VERIFY(rc == 0);
```

Comment:
Ok, this is stupid, and gives away that I haven't actually tested the Serenity version beside seeing that it boots 😀

Will submit a fix soon.